### PR TITLE
feat: add schema support for pgsql showAllTables

### PR DESCRIPTION
**Pull request by christinittyjohn**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Adds support for 
```queryInterface.showAllTables({schema: 'my_schema'})```

override showAllTables queryInterface for postgresql.

Updated the postgresql queryGenerator to use the schema passed from queryInterface#showAllTables

Closes https://github.com/sequelize/sequelize/issues/11948

### CODE REVIEW

File: packages/core/test/unit/query-generator/show-tables-query.test.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+const queryGeneratorWithSchema
+  = sequelizeWithSchema.getQueryInterface().queryGenerator;
+
+describe('QueryGenerator#showTablesQuery', () => {
+  it('should use the schema from initialization options', async () => {
+    expectsql(queryGeneratorWithSchema.showTablesQuery(), {
+      postgres: `SELECT table_name FROM information_schema.tables WHERE table_schema = 'schema_at_init' AND table_type LIKE '%TABLE' AND table_name != 'spatial_ref_sys';`,
+      mssql:
+        'SELECT TABLE_NAME, TABLE_SCHEMA FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = \'BASE TABLE\';',
+      // TODO fix TABLE_SCHEM typo and missing quotes for CURRENT SCHEMA
```
Comment (WikiRik) [inline packages/core/test/unit/query-generator/show-tables-query.test.ts:18]: One quick comment on this; I believe that this is actually correct. `TABLE_SCHEM` does give results when looking through the documentation and `CURRENT SCHEMA` is also valid since it's not a specific schema name but a keyword

Comment (christinittyjohn) [reply packages/core/test/unit/query-generator/show-tables-query.test.ts]: Thank you 🙂 

Removed the TODO comment

File: packages/core/src/dialects/abstract/query-generator.d.ts [hunk 2]
```diff
@@ -214,4 +214,6 @@ export class AbstractQueryGenerator extends AbstractQueryGeneratorTypeScript {
    * @param bind A mutable object to which bind parameters will be added.
    */
   bindParam(bind: Record<string, unknown>): (newBind: unknown) => string;
+  showTablesQuery(schemaName?: string): string;
+  showTablesQuery(database: string, options: EscapeOptions): string;
```
Comment (ephys) [inline packages/core/src/dialects/abstract/query-generator.d.ts:218]: this second signature is not necessary, `database` is a misnomer as it's actually the schema too. Snowflake accepts an options bag but it's not necessary and can be removed

File: packages/core/src/dialects/postgres/query-generator.d.ts [hunk 3]
```diff
@@ -7,4 +7,6 @@ type PgEnumNameOptions = {
 
 export class PostgresQueryGenerator extends PostgresQueryGeneratorTypeScript {
   pgEnumName(tableName: TableName, columnName: string, options?: PgEnumNameOptions): string;
+
+  showTablesQuery(showTablesSchema?: string): string;
```
Comment (ephys) [inline packages/core/src/dialects/postgres/query-generator.d.ts:11]: you don't need to repeat it here, it will be inherited from query-generator

File: packages/core/src/dialects/postgres/query-interface.d.ts [hunk 4]
```diff
@@ -1,9 +1,20 @@
-import type { Sequelize } from '../../sequelize.js';
+import type { QueryRawOptions, Sequelize } from '../../sequelize.js';
 import { AbstractQueryInterface } from '../abstract/query-interface.js';
 import type { PostgresQueryGenerator } from './query-generator.js';
 
+export type SchemaOption = {
```
Comment (ephys) [inline packages/core/src/dialects/postgres/query-interface.d.ts:5]: This SchemaOption type does not seem to be used

File: packages/core/src/dialects/postgres/query-interface.d.ts [hunk 5]
```diff
@@ -1,9 +1,20 @@
...
+  // specify the schema to get the tables from
+  schema?: string,
+};
+
+export interface QueryRawOptionsWithSchema extends QueryRawOptions {
```
Comment (ephys) [inline packages/core/src/dialects/postgres/query-interface.d.ts:10]:
```suggestion
export interface ShowAllTablesOptions extends QueryRawOptions {
```

File: packages/core/src/dialects/postgres/query-interface.js [hunk 6]
```diff
@@ -260,4 +260,28 @@ export class PostgresQueryInterface extends AbstractQueryInterface {
...
+ * @override
+ */
+  async showAllTables(options) {
+    options = {
+      ...options,
+      raw: true,
+      type: QueryTypes.SHOWTABLES,
+    };
+
+    const showTablesSql = this.queryGenerator.showTablesQuery(options.schema);
```
Comment (ephys) [inline packages/core/src/dialects/postgres/query-interface.js:282]: this shouldn't be put at the postgres level, it should have the same behavior across all dialects so it should go in AbstractQueryInterface

**Review by ephys — CHANGES_REQUESTED** [review]

**Comment by christinittyjohn** [comment]
Hi @ephys ,

After this change,  `dropAllTables` is not working for Postgres. 
This is happening because:
1. `this.sequelize.config.database` is used as an argument for `showTablesQuery` from `AbstractQueryInterface.showAllTables`.
2. `PostgresQueryGenerator.showTablesQuery` has started taking schema as input instead of using Sequelize options.schema OR `public` - [ref](https://github.com/sequelize/sequelize/pull/15668/files#diff-e3b0efdaf58b9d143317b9622917e60f2c585efdba39d5d7dc52e706d4469766R170).

I can see the following Solutions :
A -  Add support for schema in `dropAllTables`.
OR
B - guard against schema values that equal to`this.sequelize.config.database` in `PostgresQueryGenerator.showTablesQuery` so that it uses options.schema or public like it used to. Will this cause issues when the schema used is same as the database name ?

Are there any other ways to fix this ? If not which approach should I go with ?

**Comment by ephys** [comment]
You can use `dialect.getDefaultSchema()` instead of `this.sequelize.config.database`, it should return the proper schema per dialect

**Comment by christinittyjohn** [comment]
Used `getDefaultSchema` and the instance schema option as fallback values in `AbstractQueryInterface.showAllTables` @ephys 

https://github.com/sequelize/sequelize/pull/15668/commits/6abf91bb109a3f949016f1011c5b972951228044

File: packages/core/src/dialects/postgres/query-interface.js (hunk 6 above)
Comment (christinittyjohn) [reply packages/core/src/dialects/postgres/query-interface.js]: Updated abstract showAllTables instead.

File: packages/core/src/dialects/postgres/query-interface.d.ts (hunk 4 above)
Comment (christinittyjohn) [reply packages/core/src/dialects/postgres/query-interface.d.ts]: Removed this.

File: packages/core/src/dialects/postgres/query-generator.d.ts (hunk 3 above)
Comment (christinittyjohn) [reply packages/core/src/dialects/postgres/query-generator.d.ts]: Removed this.

File: packages/core/src/dialects/abstract/query-generator.d.ts (hunk 2 above)
Comment (christinittyjohn) [reply packages/core/src/dialects/abstract/query-generator.d.ts]: Removed this.

File: packages/core/test/integration/query-interface/show-all-tables.test.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    await _sequelize.queryInterface.dropAllTables();
+    if (!dialect.supports.schemas) {
+      return;
+    }
+
+    await Promise.all(['schema_1', 'schema_2', 'schema_3', 'schema_4'].map(async schema => _sequelize.dropSchema(schema)));
+  });
+
+  const getSequelizeInstanceWithSchema = async () => {
+    const _sequelize = createSequelizeInstance({ schema: 'schema_3' });
```
Comment (ephys) [inline packages/core/test/integration/query-interface/show-all-tables.test.ts:24]: We already use underscore prefixes to mute "unused vars" errors, prefer a different name

**Review by ephys — CHANGES_REQUESTED** [review]
Thanks for updating! The implementation looks good but I have a few notes on the test side

File: packages/core/test/unit/query-generator/show-tables-query.test.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        mariadb: `SELECT TABLE_NAME, TABLE_SCHEMA FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE'  AND TABLE_SCHEMA = 'schema_not_at_init';`,
+        sqlite:
+          'SELECT name FROM `sqlite_master` WHERE type=\'table\' and name!=\'sqlite_sequence\';',
+        snowflake:
+          'SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = \'BASE TABLE\' AND TABLE_SCHEMA = \'schema_not_at_init\';',
+        db2: 'SELECT TABNAME AS "tableName", TRIM(TABSCHEMA) AS "tableSchema" FROM SYSCAT.TABLES WHERE TABSCHEMA = USER AND TYPE = \'T\' ORDER BY TABSCHEMA, TABNAME',
+      },
+    );
+
+    expectsql(queryGenerator.showTablesQuery(
```
[the remark this answers is not in the document]
Comment (christinittyjohn) [reply packages/core/test/unit/query-generator/show-tables-query.test.ts:46]: the queryGenerator used here belongs to a Sequelize instance that does not have a schema while `queryGeneratorWithSchema ` does.

[the remark this answers is not in the document]
Comment (christinittyjohn) [reply packages/core/test/unit/query-generator/show-tables-query.test.ts:46]: Yeah, will remove it.

File: packages/core/test/integration/query-interface/show-all-tables.test.ts [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+          type: QueryTypes.SHOWTABLES,
+        },
+      ));
+      await sequelize.query('DROP DATABASE my_test_db');
+      expect(tableNames).to.deep.equal(['my_test_table1', 'my_test_table2']);
+    });
+  }
+
+  describe('schema support tests', () => {
+    if (!dialect.supports.schemas || dialectName !== 'postgres') {
```
[the remark this answers is not in the document]
Comment (christinittyjohn) [reply packages/core/test/integration/query-interface/show-all-tables.test.ts:143]: The assertion will fail if we run this for non postgresql dialects that do not support schema in showAllTables.

File: packages/core/test/integration/query-interface/show-all-tables.test.ts [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    return queryResults.map(normalizeTableNames);
+  };
+
+  it('should not contain views', async () => {
+    async function cleanup(_sequelize: Sequelize) {
+      if (dialectName === 'db2') {
+        // DB2 does not support DROP VIEW IF EXISTS
+        try {
+          await _sequelize.query('DROP VIEW V_Fail');
+        } catch (error: any) {
```
[the remark this answers is not in the document]
Comment (christinittyjohn) [reply packages/core/test/integration/query-interface/show-all-tables.test.ts:78]: This causes a type error for `error.cause.sqlcode`

**Comment by ephys** [comment]
Sorry for the delay
There is a very minor conflict, could you fix it then re-request a review?

File: packages/core/test/unit/utils/attribute-syntax.test.ts [hunk 11]
```diff
@@ -1,7 +1,7 @@
-import { expect } from 'chai';
```
Comment (WikiRik) [inline packages/core/test/unit/utils/attribute-syntax.test.ts:1]: You can undo the changes in this file and then the linter passes again

File: packages/core/src/dialects/abstract/query-generator.d.ts [hunk 12]
```diff
@@ -93,6 +94,11 @@ export interface RemoveColumnQueryOptions {
   ifExists?: boolean;
 }
 
+export interface ShowAllTablesOptions extends QueryRawOptions {
+  // specify the schema to get the tables from
```
Comment (WikiRik) [inline packages/core/src/dialects/abstract/query-generator.d.ts:98]:
```suggestion
  // specify the schema to get the tables from. Only supported by postgres
```

File: packages/core/src/dialects/abstract/query-generator.d.ts [hunk 13]
```diff
@@ -93,6 +94,11 @@ export interface RemoveColumnQueryOptions {
   ifExists?: boolean;
 }
 
+export interface ShowAllTablesOptions extends QueryRawOptions {
```
Comment (WikiRik) [inline packages/core/src/dialects/abstract/query-generator.d.ts:97]: Shouldn't this belong in `query-interface.d.ts` since it's not used for `showTablesQuery`?

File: packages/core/src/dialects/postgres/query-generator.js [hunk 14]
```diff
@@ -167,8 +167,8 @@ export class PostgresQueryGenerator extends PostgresQueryGeneratorTypeScript {
     return `DROP TABLE IF EXISTS ${this.quoteTable(tableName)}${options?.cascade ? ' CASCADE' : ''};`;
   }
 
-  showTablesQuery() {
-    const schema = this.options.schema || 'public';
+  showTablesQuery(showTableSchema) {
+    const schema = showTableSchema || this.options.schema || 'public';
```
Comment (WikiRik) [inline packages/core/src/dialects/postgres/query-generator.js:172]:
```suggestion
  showTablesQuery(schema) {
```

You are passing the correct schema in `QueryInterface.showAllTables` already

File: packages/core/test/integration/query-interface/show-all-tables.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (WikiRik) [reply packages/core/test/integration/query-interface/show-all-tables.test.ts:140]: Can you add a TODO that support for other dialects should be added? Or you can add the support for it in this PR directly, I'm fine with both

[the remark this answers is not in the document]
Comment (christinittyjohn) [reply packages/core/test/integration/query-interface/show-all-tables.test.ts:140]: I'll try to add schema in showAllTables support for DB2 & MSSQL other Dialects seem to support schema in the showTablesQuery.

The plan is to update:

DB2
1. Update the [Query generator](https://github.com/sequelize/sequelize/blob/main/packages/core/src/dialects/db2/query-generator.js#L190) to use schema argument
2. Since `getDefaultSchema` is not defined for DB2, use USER as the fallback value.

MSSQL
1. Update the [Query generator](https://github.com/sequelize/sequelize/blob/main/packages/core/src/dialects/mssql/query-generator.js#L237) to use schema argument which will be used as the [TABLE_SCHEMA](https://learn.microsoft.com/en-us/sql/relational-databases/system-information-schema-views/columns-transact-sql?view=sql-server-ver16)

File: packages/core/src/dialects/abstract/query-generator.d.ts (hunk 12 above)
Comment (christinittyjohn) [reply packages/core/src/dialects/abstract/query-generator.d.ts]: I'm updating the dialects to support Schema.

File: packages/core/src/dialects/abstract/query-generator.d.ts (hunk 13 above)
Comment (christinittyjohn) [reply packages/core/src/dialects/abstract/query-generator.d.ts]: Yes, moved `ShowAllTablesOptions` into `query-interface.d.ts`.

File: packages/core/src/dialects/postgres/query-generator.js (hunk 14 above)
Comment (christinittyjohn) [reply packages/core/src/dialects/postgres/query-generator.js:172]: Yup, removed this. 

'public' is defined as the default schema in `getDefaultSchema` for postgres.

**Comment by ephys** [comment]
I updated the branch to see what the failing unit tests were

**Review by ephys — CHANGES_REQUESTED** [review]
Will re-review once tests are fixed :)

**Comment by WikiRik** [comment]
Closing this, since the feature has been added in #16452
